File: jscomp/ext/bal_tree.ml
(***********************************************************************)
(*                                                                     *)
(*                                OCaml                                *)
(*                                                                     *)
(*            Xavier Leroy, projet Cristal, INRIA Rocquencourt         *)
(*                                                                     *)
(*  Copyright 1996 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  under the terms of the GNU Library General Public License, with    *)
(*  the special exception on linking described in file ../LICENSE.     *)
(*                                                                     *)
(***********************************************************************)
(** balanced tree based on stdlib distribution *)

type 'a t  = 'a Bal_set_common.t 
open Bal_set_common


(* Splitting.  
    split x s returns a triple (l, present, r) where
        - l is the set of elements of s that are < x
        - r is the set of elements of s that are > x
        - present is false if s contains no element equal to x,
          or true if s contains an element equal to x.
*)

let rec split x (tree : _ t ) : _ t  *  bool * _ t =
  match tree with 
  | Empty ->
    (Empty, false, Empty)
  | Node(l, v, r, _) ->
    let c = Pervasives.compare x v in
    if c = 0 then (l, true, r)
    else if c < 0 then
      let (ll, pres, rl) = split x l in (ll, pres, internal_join rl v r)
    else
      let (lr, pres, rr) = split x r in (internal_join l v lr, pres, rr)

let rec add x (tree : _ t) : _ t  =
  match tree with 
  | Empty -> Node(Empty, x, Empty, 1)
  | Node(l, v, r, _) as t ->
    let c = Pervasives.compare x v in
    if c = 0 then t else
    if c < 0 then internal_bal (add x l) v r else internal_bal l v (add x r)

let rec union (s1 : _ t) (s2 : _ t) =
  match (s1, s2) with
  | (Empty, t2) -> t2
  | (t1, Empty) -> t1
  | (Node(l1, v1, r1, h1), Node(l2, v2, r2, h2)) ->
    if h1 >= h2 then
      if h2 = 1 then add v2 s1 else begin
        let (l2, _, r2) = split v1 s2 in
        internal_join (union l1 l2) v1 (union r1 r2)
      end
    else
    if h1 = 1 then add v1 s2 else begin
      let (l1, _, r1) = split v2 s1 in
      internal_join (union l1 l2) v2 (union r1 r2)
    end    

let rec inter s1 s2 =
  match (s1, s2) with
  | (Empty, t2) -> Empty
  | (t1, Empty) -> Empty
  | (Node(l1, v1, r1, _), t2) ->
    begin match split v1 t2 with
      | (l2, false, r2) ->
        internal_concat (inter l1 l2) (inter r1 r2)
      | (l2, true, r2) ->
        internal_join (inter l1 l2) v1 (inter r1 r2)
    end 

let rec diff s1 s2 =
  match (s1, s2) with
  | (Empty, t2) -> Empty
  | (t1, Empty) -> t1
  | (Node(l1, v1, r1, _), t2) ->
    begin match split v1 t2 with
      | (l2, false, r2) ->
        internal_join (diff l1 l2) v1 (diff r1 r2)
      | (l2, true, r2) ->
        internal_concat (diff l1 l2) (diff r1 r2)    
    end


let rec mem x (tree : _ t) =
  match tree with 
  | Empty -> false
  | Node(l, v, r, _) ->
    let c = Pervasives.compare x v in
    c = 0 || mem x (if c < 0 then l else r)

let rec remove x = function
    Empty -> Empty
  | Node(l, v, r, _) ->
    let c = Pervasives.compare x v in
    if c = 0 then internal_merge l r else
    if c < 0 then internal_bal (remove x l) v r else internal_bal l v (remove x r)


let rec subset (s1 : _ t) (s2 : _ t) =
  match (s1, s2) with
  | Empty, _ -> true
  | _, Empty -> false
  | Node (l1, v1, r1, _), (Node (l2, v2, r2, _) as t2) ->
    let c = Pervasives.compare v1 v2 in
    if c = 0 then
      subset l1 l2 && subset r1 r2
    else if c < 0 then
      subset (Node (l1, v1, Empty, 0)) l2 && subset r1 t2
    else
      subset (Node (Empty, v1, r1, 0)) r2 && subset l1 t2

let compare s1 s2 = Bal_set_common.compare Pervasives.compare s1 s2 

let equal s1 s2 = compare s1 s2 = 0

let rec find x = function
    Empty -> raise Not_found
  | Node(l, v, r, _) ->
    let c = Pervasives.compare x v in
    if c = 0 then v
    else find x (if c < 0 then l else r)



let of_list l =
  match l with
  | [] -> empty
  | [x0] -> singleton x0
  | [x0; x1] -> add x1 (singleton x0)
  | [x0; x1; x2] -> add x2 (add x1 (singleton x0))
  | [x0; x1; x2; x3] -> add x3 (add x2 (add x1 (singleton x0)))
  | [x0; x1; x2; x3; x4] -> add x4 (add x3 (add x2 (add x1 (singleton x0))))
  | _ -> of_sorted_list (List.sort_uniq Pervasives.compare l)

let of_array l = 
  Array.fold_left (fun  acc x -> add x acc) empty l




let invariant t =
  Bal_set_common.invariant Pervasives.compare t 

module Make ( S : Set.OrderedType) = struct
  type elt = S.t
  type nonrec t = elt t 
  let empty = Bal_set_common.empty 
  let is_empty = Bal_set_common.is_empty
  let iter = Bal_set_common.iter 
  let fold = Bal_set_common.fold 
  let for_all = Bal_set_common.for_all 
  let exists = Bal_set_common.exists 
  let singleton = Bal_set_common.singleton 
  let cardinal = Bal_set_common.cardinal
  let elements = Bal_set_common.elements
  let min_elt = Bal_set_common.min_elt
  let max_elt = Bal_set_common.max_elt
  let choose = Bal_set_common.choose 
  let of_sorted_list = Bal_set_common.of_sorted_list
  let of_sorted_array = Bal_set_common.of_sorted_array
  let partition = Bal_set_common.partition 
  let filter = Bal_set_common.filter 
  let of_sorted_list = Bal_set_common.of_sorted_list
  let of_sorted_array = Bal_set_common.of_sorted_array
  let rec split x = function
    | Empty ->
      (Empty, false, Empty)
    | Node(l, v, r, _) ->
      let c = S.compare x v in
      if c = 0 then (l, true, r)
      else if c < 0 then
        let (ll, pres, rl) = split x l in (ll, pres, internal_join rl v r)
      else
        let (lr, pres, rr) = split x r in (internal_join l v lr, pres, rr)
  let rec add x = function
      Empty -> Node(Empty, x, Empty, 1)
    | Node(l, v, r, _) as t ->
      let c = S.compare x v in
      if c = 0 then t else
      if c < 0 then internal_bal (add x l) v r else internal_bal l v (add x r)

  let rec union s1 s2 =
    match (s1, s2) with
    | (Empty, t2) -> t2
    | (t1, Empty) -> t1
    | (Node(l1, v1, r1, h1), Node(l2, v2, r2, h2)) ->
      if h1 >= h2 then
        if h2 = 1 then add v2 s1 else begin
          let (l2, _, r2) = split v1 s2 in
          internal_join (union l1 l2) v1 (union r1 r2)
        end
      else
      if h1 = 1 then add v1 s2 else begin
        let (l1, _, r1) = split v2 s1 in
        internal_join (union l1 l2) v2 (union r1 r2)
      end    

  let rec inter s1 s2 =
    match (s1, s2) with
    | (Empty, t2) -> Empty
    | (t1, Empty) -> Empty
    | (Node(l1, v1, r1, _), t2) ->
      begin match split v1 t2 with
        | (l2, false, r2) ->
          internal_concat (inter l1 l2) (inter r1 r2)
        | (l2, true, r2) ->
          internal_join (inter l1 l2) v1 (inter r1 r2)
      end 

  let rec diff s1 s2 =
    match (s1, s2) with
    | (Empty, t2) -> Empty
    | (t1, Empty) -> t1
    | (Node(l1, v1, r1, _), t2) ->
      begin match split v1 t2 with
        | (l2, false, r2) ->
          internal_join (diff l1 l2) v1 (diff r1 r2)
        | (l2, true, r2) ->
          internal_concat (diff l1 l2) (diff r1 r2)    
      end


  let rec mem x = function
      Empty -> false
    | Node(l, v, r, _) ->
      let c = S.compare x v in
      c = 0 || mem x (if c < 0 then l else r)

  let rec remove x = function
      Empty -> Empty
    | Node(l, v, r, _) ->
      let c = S.compare x v in
      if c = 0 then internal_merge l r else
      if c < 0 then internal_bal (remove x l) v r else internal_bal l v (remove x r)

  let compare s1 s2 = Bal_set_common.compare Pervasives.compare s1 s2 


  let equal s1 s2 =
    compare s1 s2 = 0

  let rec subset s1 s2 =
    match (s1, s2) with
      Empty, _ ->
      true
    | _, Empty ->
      false
    | Node (l1, v1, r1, _), (Node (l2, v2, r2, _) as t2) ->
      let c = S.compare v1 v2 in
      if c = 0 then
        subset l1 l2 && subset r1 r2
      else if c < 0 then
        subset (Node (l1, v1, Empty, 0)) l2 && subset r1 t2
      else
        subset (Node (Empty, v1, r1, 0)) r2 && subset l1 t2




  let rec find x = function
      Empty -> raise Not_found
    | Node(l, v, r, _) ->
      let c = S.compare x v in
      if c = 0 then v
      else find x (if c < 0 then l else r)



  let of_list l =
    match l with
    | [] -> empty
    | [x0] -> singleton x0
    | [x0; x1] -> add x1 (singleton x0)
    | [x0; x1; x2] -> add x2 (add x1 (singleton x0))
    | [x0; x1; x2; x3] -> add x3 (add x2 (add x1 (singleton x0)))
    | [x0; x1; x2; x3; x4] -> add x4 (add x3 (add x2 (add x1 (singleton x0))))
    | _ -> of_sorted_list (List.sort_uniq S.compare l)

  let of_array l = 
    Array.fold_left (fun  acc x -> add x acc) empty l

  (* also check order *)
  let invariant t =
    check t ;
    is_ordered S.compare t          
end 

File: jscomp/test/bal_tree.js
'use strict';

var Caml_obj                = require("../../lib/js/caml_obj");
var Caml_builtin_exceptions = require("../../lib/js/caml_builtin_exceptions");
var Bal_set_common          = require("./bal_set_common");
var Curry                   = require("../../lib/js/curry");
var $$Array                 = require("../../lib/js/array");
var List                    = require("../../lib/js/list");

function split(x, tree) {
  if (tree) {
    var r = tree[2];
    var v = tree[1];
    var l = tree[0];
    var c = Caml_obj.caml_compare(x, v);
    if (c) {
      if (c < 0) {
        var match = split(x, l);
        return /* tuple */[
                match[0],
                match[1],
                Bal_set_common.internal_join(match[2], v, r)
              ];
      }
      else {
        var match$1 = split(x, r);
        return /* tuple */[
                Bal_set_common.internal_join(l, v, match$1[0]),
                match$1[1],
                match$1[2]
              ];
      }
    }
    else {
      return /* tuple */[
              l,
              /* true */1,
              r
            ];
    }
  }
  else {
    return /* tuple */[
            /* Empty */0,
            /* false */0,
            /* Empty */0
          ];
  }
}

function add(x, tree) {
  if (tree) {
    var r = tree[2];
    var v = tree[1];
    var l = tree[0];
    var c = Caml_obj.caml_compare(x, v);
    if (c) {
      if (c < 0) {
        return Bal_set_common.internal_bal(add(x, l), v, r);
      }
      else {
        return Bal_set_common.internal_bal(l, v, add(x, r));
      }
    }
    else {
      return tree;
    }
  }
  else {
    return /* Node */[
            /* Empty */0,
            x,
            /* Empty */0,
            1
          ];
  }
}

function union(s1, s2) {
  if (s1) {
    if (s2) {
      var h2 = s2[3];
      var v2 = s2[1];
      var h1 = s1[3];
      var v1 = s1[1];
      if (h1 >= h2) {
        if (h2 === 1) {
          return add(v2, s1);
        }
        else {
          var match = split(v1, s2);
          return Bal_set_common.internal_join(union(s1[0], match[0]), v1, union(s1[2], match[2]));
        }
      }
      else if (h1 === 1) {
        return add(v1, s2);
      }
      else {
        var match$1 = split(v2, s1);
        return Bal_set_common.internal_join(union(match$1[0], s2[0]), v2, union(match$1[2], s2[2]));
      }
    }
    else {
      return s1;
    }
  }
  else {
    return s2;
  }
}

function inter(s1, s2) {
  if (s1) {
    if (s2) {
      var r1 = s1[2];
      var v1 = s1[1];
      var l1 = s1[0];
      var match = split(v1, s2);
      var l2 = match[0];
      if (match[1] !== 0) {
        return Bal_set_common.internal_join(inter(l1, l2), v1, inter(r1, match[2]));
      }
      else {
        return Bal_set_common.internal_concat(inter(l1, l2), inter(r1, match[2]));
      }
    }
    else {
      return /* Empty */0;
    }
  }
  else {
    return /* Empty */0;
  }
}

function diff(s1, s2) {
  if (s1) {
    if (s2) {
      var r1 = s1[2];
      var v1 = s1[1];
      var l1 = s1[0];
      var match = split(v1, s2);
      var l2 = match[0];
      if (match[1] !== 0) {
        return Bal_set_common.internal_concat(diff(l1, l2), diff(r1, match[2]));
      }
      else {
        return Bal_set_common.internal_join(diff(l1, l2), v1, diff(r1, match[2]));
      }
    }
    else {
      return s1;
    }
  }
  else {
    return /* Empty */0;
  }
}

function mem(x, _tree) {
  while(true) {
    var tree = _tree;
    if (tree) {
      var c = Caml_obj.caml_compare(x, tree[1]);
      if (c) {
        _tree = c < 0 ? tree[0] : tree[2];
        continue ;
        
      }
      else {
        return /* true */1;
      }
    }
    else {
      return /* false */0;
    }
  };
}

function remove(x, param) {
  if (param) {
    var r = param[2];
    var v = param[1];
    var l = param[0];
    var c = Caml_obj.caml_compare(x, v);
    if (c) {
      if (c < 0) {
        return Bal_set_common.internal_bal(remove(x, l), v, r);
      }
      else {
        return Bal_set_common.internal_bal(l, v, remove(x, r));
      }
    }
    else {
      return Bal_set_common.internal_merge(l, r);
    }
  }
  else {
    return /* Empty */0;
  }
}

function subset(_s1, _s2) {
  while(true) {
    var s2 = _s2;
    var s1 = _s1;
    if (s1) {
      if (s2) {
        var r2 = s2[2];
        var l2 = s2[0];
        var r1 = s1[2];
        var v1 = s1[1];
        var l1 = s1[0];
        var c = Caml_obj.caml_compare(v1, s2[1]);
        if (c) {
          if (c < 0) {
            if (subset(/* Node */[
                    l1,
                    v1,
                    /* Empty */0,
                    0
                  ], l2)) {
              _s1 = r1;
              continue ;
              
            }
            else {
              return /* false */0;
            }
          }
          else if (subset(/* Node */[
                  /* Empty */0,
                  v1,
                  r1,
                  0
                ], r2)) {
            _s1 = l1;
            continue ;
            
          }
          else {
            return /* false */0;
          }
        }
        else if (subset(l1, l2)) {
          _s2 = r2;
          _s1 = r1;
          continue ;
          
        }
        else {
          return /* false */0;
        }
      }
      else {
        return /* false */0;
      }
    }
    else {
      return /* true */1;
    }
  };
}

function compare(s1, s2) {
  return Bal_set_common.compare(Caml_obj.caml_compare, s1, s2);
}

function equal(s1, s2) {
  return +(compare(s1, s2) === 0);
}

function find(x, _param) {
  while(true) {
    var param = _param;
    if (param) {
      var v = param[1];
      var c = Caml_obj.caml_compare(x, v);
      if (c) {
        _param = c < 0 ? param[0] : param[2];
        continue ;
        
      }
      else {
        return v;
      }
    }
    else {
      throw Caml_builtin_exceptions.not_found;
    }
  };
}

function of_list(l) {
  if (l) {
    var match = l[1];
    var x0 = l[0];
    if (match) {
      var match$1 = match[1];
      var x1 = match[0];
      if (match$1) {
        var match$2 = match$1[1];
        var x2 = match$1[0];
        if (match$2) {
          var match$3 = match$2[1];
          var x3 = match$2[0];
          if (match$3) {
            if (match$3[1]) {
              return Bal_set_common.of_sorted_list(List.sort_uniq(Caml_obj.caml_compare, l));
            }
            else {
              return add(match$3[0], add(x3, add(x2, add(x1, Bal_set_common.singleton(x0)))));
            }
          }
          else {
            return add(x3, add(x2, add(x1, Bal_set_common.singleton(x0))));
          }
        }
        else {
          return add(x2, add(x1, Bal_set_common.singleton(x0)));
        }
      }
      else {
        return add(x1, Bal_set_common.singleton(x0));
      }
    }
    else {
      return Bal_set_common.singleton(x0);
    }
  }
  else {
    return /* Empty */0;
  }
}

function of_array(l) {
  return $$Array.fold_left(function (acc, x) {
              return add(x, acc);
            }, /* Empty */0, l);
}

function invariant(t) {
  return Bal_set_common.invariant(Caml_obj.caml_compare, t);
}

function Make(S) {
  var split = function (x, param) {
    if (param) {
      var r = param[2];
      var v = param[1];
      var l = param[0];
      var c = Curry._2(S[/* compare */0], x, v);
      if (c) {
        if (c < 0) {
          var match = split(x, l);
          return /* tuple */[
                  match[0],
                  match[1],
                  Bal_set_common.internal_join(match[2], v, r)
                ];
        }
        else {
          var match$1 = split(x, r);
          return /* tuple */[
                  Bal_set_common.internal_join(l, v, match$1[0]),
                  match$1[1],
                  match$1[2]
                ];
        }
      }
      else {
        return /* tuple */[
                l,
                /* true */1,
                r
              ];
      }
    }
    else {
      return /* tuple */[
              /* Empty */0,
              /* false */0,
              /* Empty */0
            ];
    }
  };
  var add = function (x, t) {
    if (t) {
      var r = t[2];
      var v = t[1];
      var l = t[0];
      var c = Curry._2(S[/* compare */0], x, v);
      if (c) {
        if (c < 0) {
          return Bal_set_common.internal_bal(add(x, l), v, r);
        }
        else {
          return Bal_set_common.internal_bal(l, v, add(x, r));
        }
      }
      else {
        return t;
      }
    }
    else {
      return /* Node */[
              /* Empty */0,
              x,
              /* Empty */0,
              1
            ];
    }
  };
  var union = function (s1, s2) {
    if (s1) {
      if (s2) {
        var h2 = s2[3];
        var v2 = s2[1];
        var h1 = s1[3];
        var v1 = s1[1];
        if (h1 >= h2) {
          if (h2 === 1) {
            return add(v2, s1);
          }
          else {
            var match = split(v1, s2);
            return Bal_set_common.internal_join(union(s1[0], match[0]), v1, union(s1[2], match[2]));
          }
        }
        else if (h1 === 1) {
          return add(v1, s2);
        }
        else {
          var match$1 = split(v2, s1);
          return Bal_set_common.internal_join(union(match$1[0], s2[0]), v2, union(match$1[2], s2[2]));
        }
      }
      else {
        return s1;
      }
    }
    else {
      return s2;
    }
  };
  var inter = function (s1, s2) {
    if (s1) {
      if (s2) {
        var r1 = s1[2];
        var v1 = s1[1];
        var l1 = s1[0];
        var match = split(v1, s2);
        var l2 = match[0];
        if (match[1] !== 0) {
          return Bal_set_common.internal_join(inter(l1, l2), v1, inter(r1, match[2]));
        }
        else {
          return Bal_set_common.internal_concat(inter(l1, l2), inter(r1, match[2]));
        }
      }
      else {
        return /* Empty */0;
      }
    }
    else {
      return /* Empty */0;
    }
  };
  var diff = function (s1, s2) {
    if (s1) {
      if (s2) {
        var r1 = s1[2];
        var v1 = s1[1];
        var l1 = s1[0];
        var match = split(v1, s2);
        var l2 = match[0];
        if (match[1] !== 0) {
          return Bal_set_common.internal_concat(diff(l1, l2), diff(r1, match[2]));
        }
        else {
          return Bal_set_common.internal_join(diff(l1, l2), v1, diff(r1, match[2]));
        }
      }
      else {
        return s1;
      }
    }
    else {
      return /* Empty */0;
    }
  };
  var mem = function (x, _param) {
    while(true) {
      var param = _param;
      if (param) {
        var c = Curry._2(S[/* compare */0], x, param[1]);
        if (c) {
          _param = c < 0 ? param[0] : param[2];
          continue ;
          
        }
        else {
          return /* true */1;
        }
      }
      else {
        return /* false */0;
      }
    };
  };
  var remove = function (x, param) {
    if (param) {
      var r = param[2];
      var v = param[1];
      var l = param[0];
      var c = Curry._2(S[/* compare */0], x, v);
      if (c) {
        if (c < 0) {
          return Bal_set_common.internal_bal(remove(x, l), v, r);
        }
        else {
          return Bal_set_common.internal_bal(l, v, remove(x, r));
        }
      }
      else {
        return Bal_set_common.internal_merge(l, r);
      }
    }
    else {
      return /* Empty */0;
    }
  };
  var compare = function (s1, s2) {
    return Bal_set_common.compare(Caml_obj.caml_compare, s1, s2);
  };
  var equal = function (s1, s2) {
    return +(compare(s1, s2) === 0);
  };
  var subset = function (_s1, _s2) {
    while(true) {
      var s2 = _s2;
      var s1 = _s1;
      if (s1) {
        if (s2) {
          var r2 = s2[2];
          var l2 = s2[0];
          var r1 = s1[2];
          var v1 = s1[1];
          var l1 = s1[0];
          var c = Curry._2(S[/* compare */0], v1, s2[1]);
          if (c) {
            if (c < 0) {
              if (subset(/* Node */[
                      l1,
                      v1,
                      /* Empty */0,
                      0
                    ], l2)) {
                _s1 = r1;
                continue ;
                
              }
              else {
                return /* false */0;
              }
            }
            else if (subset(/* Node */[
                    /* Empty */0,
                    v1,
                    r1,
                    0
                  ], r2)) {
              _s1 = l1;
              continue ;
              
            }
            else {
              return /* false */0;
            }
          }
          else if (subset(l1, l2)) {
            _s2 = r2;
            _s1 = r1;
            continue ;
            
          }
          else {
            return /* false */0;
          }
        }
        else {
          return /* false */0;
        }
      }
      else {
        return /* true */1;
      }
    };
  };
  var find = function (x, _param) {
    while(true) {
      var param = _param;
      if (param) {
        var v = param[1];
        var c = Curry._2(S[/* compare */0], x, v);
        if (c) {
          _param = c < 0 ? param[0] : param[2];
          continue ;
          
        }
        else {
          return v;
        }
      }
      else {
        throw Caml_builtin_exceptions.not_found;
      }
    };
  };
  var of_list = function (l) {
    if (l) {
      var match = l[1];
      var x0 = l[0];
      if (match) {
        var match$1 = match[1];
        var x1 = match[0];
        if (match$1) {
          var match$2 = match$1[1];
          var x2 = match$1[0];
          if (match$2) {
            var match$3 = match$2[1];
            var x3 = match$2[0];
            if (match$3) {
              if (match$3[1]) {
                return Bal_set_common.of_sorted_list(List.sort_uniq(S[/* compare */0], l));
              }
              else {
                return add(match$3[0], add(x3, add(x2, add(x1, Bal_set_common.singleton(x0)))));
              }
            }
            else {
              return add(x3, add(x2, add(x1, Bal_set_common.singleton(x0))));
            }
          }
          else {
            return add(x2, add(x1, Bal_set_common.singleton(x0)));
          }
        }
        else {
          return add(x1, Bal_set_common.singleton(x0));
        }
      }
      else {
        return Bal_set_common.singleton(x0);
      }
    }
    else {
      return /* Empty */0;
    }
  };
  var of_array = function (l) {
    return $$Array.fold_left(function (acc, x) {
                return add(x, acc);
              }, /* Empty */0, l);
  };
  var invariant = function (t) {
    Bal_set_common.check(t);
    return Bal_set_common.is_ordered(S[/* compare */0], t);
  };
  return /* module */[
          /* empty : Empty */0,
          /* is_empty */Bal_set_common.is_empty,
          /* iter */Bal_set_common.iter,
          /* fold */Bal_set_common.fold,
          /* for_all */Bal_set_common.for_all,
          /* exists */Bal_set_common.exists,
          /* singleton */Bal_set_common.singleton,
          /* cardinal */Bal_set_common.cardinal,
          /* elements */Bal_set_common.elements,
          /* min_elt */Bal_set_common.min_elt,
          /* max_elt */Bal_set_common.max_elt,
          /* choose */Bal_set_common.choose,
          /* partition */Bal_set_common.partition,
          /* filter */Bal_set_common.filter,
          /* of_sorted_list */Bal_set_common.of_sorted_list,
          /* of_sorted_array */Bal_set_common.of_sorted_array,
          /* split */split,
          /* add */add,
          /* union */union,
          /* inter */inter,
          /* diff */diff,
          /* mem */mem,
          /* remove */remove,
          /* compare */compare,
          /* equal */equal,
          /* subset */subset,
          /* find */find,
          /* of_list */of_list,
          /* of_array */of_array,
          /* invariant */invariant
        ];
}

exports.split     = split;
exports.add       = add;
exports.union     = union;
exports.inter     = inter;
exports.diff      = diff;
exports.mem       = mem;
exports.remove    = remove;
exports.subset    = subset;
exports.compare   = compare;
exports.equal     = equal;
exports.find      = find;
exports.of_list   = of_list;
exports.of_array  = of_array;
exports.invariant = invariant;
exports.Make      = Make;
/* No side effect */
